databuilder/__main__.py

```diff
@@ -66,113 +66,122 @@ def main(args=None):
     elif options.which == "print_help":
         parser.print_help()
     else:
-        assert False, f"Unhandler subcommand: {options.which}"
+        assert False, f"Unhandled subcommand: {options.which}"
 
 
 def build_parser():
     parser = ArgumentParser(
         prog="databuilder", description="Generate datasets in OpenSAFELY"
     )
     parser.set_defaults(which="print_help")
+
     subparsers = parser.add_subparsers(help="sub-command help")
+    add_generate_dataset(subparsers)
+    add_validate_dataset_definition(subparsers)
+    add_generate_measures(subparsers)
+    add_test_connection(subparsers)
+    add_generate_docs(subparsers)
 
-    generate_dataset_parser = subparsers.add_parser(
-        "generate_dataset", help="Generate a dataset"
-    )
-    generate_dataset_parser.set_defaults(which="generate_dataset")
-    generate_dataset_parser.add_argument(
+    return parser
+
+
+def add_generate_dataset(subparsers):
+    parser = subparsers.add_parser("generate_dataset", help="Generate a dataset")
+    parser.set_defaults(which="generate_dataset")
```

Comment:
A pattern we use in the CLI tool which avoids the big switch on `which`, is to do:
```py
parser.set_defaults(function=function_to_call)
```

And then the parsing and invocation code can just be:
```py
args = parser.parse_args()
kwargs = vars(args)
function = kwargs.pop("function")
function(**kwargs)
```

That may be distressingly magical for your taste though. I don't have any objection to doing it the more laborious way.

Reply:
No, not too magic at all. This code is so predictable and verbose that I think that sort of thing is completely justified. I just didn't take the time to work out how to do it. I'll adopt that here as well. Thanks.

Reply:
Well...

I can't see a way of doing this that doesn't make me weep (because we also take some inputs as environment variables).

The command line args and env vars aren't really consistent and coherent. So I'm going to merge this as it is and create a story for the backlog to straighten this out at some point.